### PR TITLE
Refactor and improve `FindReturnRef` to find indirect cases

### PR DESCRIPTION
**Pull request by baloghadamsoftware**
Currently, the detector `FindReturnRef` which is responisble for bugs
`EI_EXPOSE_REP`, `EI_EXPOSE_REP2`, `MS_EXPOSE_REP` and
`EI_EXPOSE_STATIC_REP2` tracks the data on the top of the stack.
However, this is not necessary in the current version of SpotBugs
because the infrastructure already does this job more accurately. This
patch removes the redundant tracking the the top of the stack and
relies on the already existing infrastructure instead. This way the
detector is able to find more true positives where returning of the
private mutable attribute happens indirectly, e.g. by first assigning it
to a local variable and then returning that local.

----

Make sure these boxes are checked before submitting your PR -- thank you!

- [x] Added an entry into `CHANGELOG.md` if you have changed SpotBugs code

### CODE REVIEW

**Comment by baloghadamsoftware** [comment]
This pull request depends on #1525 which is already accepted by @ThrawnCA but not merged yet.

**Comment by baloghadamsoftware** [comment]
#1525 is merged now. Could someone (@KengoTODA  or @ThrawnCA), please start the review on this PR?

File: spotbugsTestCases/src/java/FindReturnRefNegativeTest.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        return dateCopy;
+    }
+
+    @NoWarning("EI")
+    public Hashtable<Integer, String> getValues() {
+        return (Hashtable<Integer, String>) ht.clone();
+    }
+
+    @NoWarning("MS")
+    public static Hashtable<Integer, String> getSaticValues() {
```
Comment (ThrawnCA) [inline spotbugsTestCases/src/java/FindReturnRefNegativeTest.java:101]: Mispelled "static"

File: spotbugsTestCases/src/java/FindReturnRefNegativeTest.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
+import edu.umd.cs.findbugs.annotations.NoWarning;
+
+import java.util.Date;
+import java.util.Hashtable;
```
Comment (ThrawnCA) [inline spotbugsTestCases/src/java/FindReturnRefNegativeTest.java:4]: Why not `Map`?

**Review by ThrawnCA — CHANGES_REQUESTED** [review]

File: spotbugsTestCases/src/java/FindReturnRefNegativeTest.java (hunk 2 above)
Comment (baloghadamsoftware) [reply spotbugsTestCases/src/java/FindReturnRefNegativeTest.java]: I copied the original test from the SEI CERT rule, but now I changed it to HashMap.

**Review by ThrawnCA — DISMISSED** [review]

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindReturnRef.java [hunk 3]
```diff
@@ -152,4 +154,19 @@ private boolean isPotentialCapture(OpcodeStack.Item top) {
         return top.getRegisterNumber() != parameterCount - 1;
 
     }
+
+    private boolean isFieldOf(XField field, ClassDescriptor klass) {
```
Comment (ThrawnCA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindReturnRef.java:158]: The usual convention is "clazz", but okay.

**Review by ThrawnCA — APPROVED** [review]

**Comment by KengoTODA** [comment]
It seems that this branch contains the change from #1542 that is already merged to master.
Not sure why it happens, but I suggest you to recreate branch and PR to make the change focused on your intention.

**Pull request closed without merging**